### PR TITLE
fix(iOS): Replace uses of `CGColorRef` with UIColor to avoid manual memory management

### PR DESCRIPTION
**Pull request by Saadnajmi**
## Summary:

Update React Native on iOS to do less manual memory management, by replacing uses of `CGColorRef` with `UIColor`, and only calling `UIColor.CGColor` when needed. This results in much less manual memory management, which is probably a good thing in 2024 :D. The downside is there is a breaking change: the signature of a method in `RCTBorderDrawing` changes.

This is a followup to #46797 . After that PR merged and I tested merging React Native macOS to the `0.76-stable` branch cut commit again, I saw even more places I needed to manually call `CGColorRetain` / `CGColorRelease`. The reason is due to React Native macOS specifics (explained below) and I could update React Native macOS to not need these changes, but I thought I would at least throw up a PR to propose the changes, as it may be good to move away from using Core Graphics' C base API as much as possible.

## Longer Explanation

With https://github.com/microsoft/react-native-macos/pull/2209 , I wrote a shim of [UIGraphicsImageRenderer](https://developer.apple.com/documentation/uikit/uigraphicsimagerenderer) for macOS. The main difference is my shim calls the NSImage API [imageWithSize:flipped:drawingHandler:](https://developer.apple.com/documentation/appkit/nsimage/1519860-imagewithsize?language=objc) to shim the UIGraphicsImageRenderer API [imageWithData:](https://developer.apple.com/documentation/uikit/uiimage/1624137-imagewithdata). The difference between the two is that the macOS API copies the block, and executes it when Appkit is about to draw the image, while the iOS API executes the block right away. Because of this, I think I am hitting way more places where `CGColorRef` variables needed to be retained / released. Additionally, I hit more of them when I merge to the 0.76 branch cut commit (this is why I didn't catch it with #46797). 

Given this constraint, I have a couple of options:
1. Refactor my macOS shim to use the deprecated API [[NSImage lockFocus]](https://developer.apple.com/documentation/appkit/nsimage/1519891-lockfocus)
    - I am not a fan of this because `lockFocus` was deprecated for `imageWithSize:flipped:drawingHandler:`
2. Refactor my macOS shim to do what we used to do: Create a CGContext manually and write it to an image
    - This is probably OK. Relies on less Appkit specifics, and potentially gives more control to RN on the rendering layer, which I recall @lenaic told me is better for Fabric)
3. Refactor React Native to avoid CGColorRef altogether, and use `UIColor` (which ARC will memory manage for us) as much as possible.
    - This is the approach of this PR and my preferred approach. The downside is this changes the signature of some of the methods in `RCTBorderDrawing` which is a breaking change. I've seen other PRs do this, so maybe its OK?

## Changelog:

[IOS] [BREAKING] -  Replace uses of `CGColorRef` with UIColor to avoid manual memory management

## Test Plan:

Launching RNTester's View example (which tests a lot of the border / outline / shadow rendering) does not crash for me on both iOS and macOS.

### CODE REVIEW

File: packages/react-native/React/Views/RCTView.m [hunk 1]
```diff
@@ -807,28 +807,30 @@ - (void)displayLayer:(CALayer *)layer
...
-      // the content. For this reason, only use iOS border drawing when clipping
-      // or when the border is hidden.
+    // iOS draws borders in front of the content whereas CSS draws them behind
+    // the content. For this reason, only use iOS border drawing when clipping
+    // or when the border is hidden.
 
-      (borderInsets.top == 0 || (borderColors.top && CGColorGetAlpha(borderColors.top) == 0) || self.clipsToBounds);
+    (borderInsets.top == 0 ||
+      (borderColors.top && CGColorGetAlpha(borderColors.top.CGColor) == 0)
+      || self.clipsToBounds);
```
Comment (Saadnajmi) [inline packages/react-native/React/Views/RCTView.m:821]:
```suggestion
      (borderColors.top && CGColorGetAlpha(borderColors.top.CGColor) == 0) ||
      self.clipsToBounds);
```

**Comment by joevilches** [comment]
Hey @Saadnajmi, this looks good to me, thanks for taking this on!

**Comment by joevilches** [comment]
@Saadnajmi there are some build errors internally that coincide with the warnings about incompatible pointer on RCTView.m, mind addressing that? (I would comment on the lines but github is not letting me for some reason -_-)

**Comment by Saadnajmi** [comment]
> @Saadnajmi there are some build errors internally that coincide with the warnings about incompatible pointer on RCTView.m, mind addressing that? (I would comment on the lines but github is not letting me for some reason -_-)

Eh, I'm not sure I can see those errors, but I'll take a second look. I may have just been surprising them.

**Comment by joevilches** [comment]
@Saadnajmi GitHub added warnings here that coincide with the error. Screenshotting since I still cannot add inline comment
Screenshot 2024-10-07 at 15 01 31

**Comment by Saadnajmi** [comment]
@joevilches OK I see them now, and they are fixed. Totally just missed / ignored those, thanks!

**Pull request closed without merging**